src/v/cluster/topic_recovery_validator.cc

```diff
@@ -0,0 +1,330 @@
+// Copyright 2024 Redpanda Data, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+
+#include "topic_recovery_validator.h"
+
+#include "cloud_storage/anomalies_detector.h"
+#include "cluster/logger.h"
+
+#include <seastar/coroutine/as_future.hh>
+
+namespace cluster {
+
+class partition_validator {
+public:
+    // Each partition gets a separate retry_chain_node and a logger tied to it.
+    // From this a common retry_chain_logger is created
```

Comment:
> Do you think it's worth unit testing this partition validator on its own? Defining it in the .cc file seems to discourage testing

good point. @andijcr we want unit testing to be the first line of defense for testing new code.

however, if we have some testing that we feel like will be exercising this, i'm ok if we do this as a follow up rather than continuing to iterate on this large PR. if we don't think any of this code is getting exercised, then we need to do that.

Reply:
tests/rptest/tests/topic_recovery_test.py::TopicRecoveryTest.[test_prevent_topic_recovery|test_many_partitions] are written to test this code. i'll get a unit test going like for anomalies_detector. the dimension are [validation mode, cardinality 4], [partition damage, cardinality ~6], [download results, cardinatily ~3]